source/postprocess/stokes_residual.cc

```diff
@@ -81,18 +81,58 @@ namespace aspect
     }
 
     template <int dim>
-    void StokesResidual<dim>::stokes_solver_callback (const SimulatorAccess<dim> &sim,
-                                                      const bool /*success*/,
-                                                      const std::vector<double> &history)
+    void StokesResidual<dim>::stokes_solver_callback (const SolverControl &solver_control_cheap,
+                                                      const SolverControl &solver_control_expensive)
     {
+
       unsigned int current_solve_index = 0;
-      if (entries.size()>0 && entries.back().time == sim.get_time())
+      if (entries.size()>0 && entries.back().time == this->get_time())
         current_solve_index = entries.back().solve_index+1;
 
       DataPoint data_point;
-      data_point.time = sim.get_time();
+      data_point.time = this->get_time();
       data_point.solve_index = current_solve_index;
-      data_point.values = history;
+
+#if DEAL_II_VERSION_GTE(9,0,0)
+      data_point.values = solver_control_cheap.get_history_data();
+      if ((solver_control_cheap.last_check() == SolverControl::State::failure)
+          && (solver_control_cheap.last_step() == solver_control_cheap.max_steps()))
+        {
+          data_point.values.push_back(-1.0);
+          data_point.values.insert(data_point.values.end(),
+                                   solver_control_expensive.get_history_data().begin(),
+                                   solver_control_expensive.get_history_data().end());
+        }
+#else
+      const ExtendedSolverControl &cheap = dynamic_cast<const ExtendedSolverControl &> (solver_control_cheap);
+      const ExtendedSolverControl &expensive = dynamic_cast<const ExtendedSolverControl &> (solver_control_expensive);
+      data_point.values = cheap.get_history_data();
+      if ((solver_control_cheap.last_check() == SolverControl::State::failure)
+          && (solver_control_cheap.last_step() == solver_control_cheap.max_steps()))
+        {
+          data_point.values.push_back(-1.0);
+          data_point.values.insert(data_point.values.end(),
+                                   expensive.get_history_data().begin(),
+                                   expensive.get_history_data().end());
+        }
+
+
+      // Pre deal.II 9.0 history_data contained 0 for all iterations
+      // up to max steps (e.g. because the solver converged earlier).
+      // Remove those entries.
+      std::vector<double>::iterator zero_value = std::find(data_point.values.begin(),
+                                                           data_point.values.end(),
+                                                           0);
+      while (zero_value != data_point.values.end())
+        {
+          data_point.values.erase(zero_value);
```

Comment:
you can make that cheaper by calling `data_points.erase (zero_value, data_points.end())`.